### ISSUE TITLE
GamePad issues

### ISSUE THREAD

**Issue by endimilojkoski** [mono/MonoGame#4948]
1. Connecting X360 Gamepad while game is running the game crashes. It tries to add deviceId=0 key which already exist and I have only 1 gamepad connected.
![image](https://cloud.githubusercontent.com/assets/13436441/16155581/5054cae6-34b0-11e6-8f86-0ba0b505bd23.png)

2. When starting the game with 1 X360 and 2 generic gamepads the game starts fine and the X360 gamepad works but the 2 generic ones are not recognised.
In this scenario I see 2 GamePads and 4 Joysticks, each have +1 unknown device. So GamePad gets only the X360 and Joystick get all gamepads connected. This is where they are assigned:
![image](https://cloud.githubusercontent.com/assets/13436441/16155687/c960f00e-34b0-11e6-801c-6b95f7fa36fd.png)

3. On my old MacBook the X360 gamepad works fine if connected before starting the game but for a friend of mine with newer MacBook the game crashes if his X360 gamepad is connected. This is already reported in issue #4932 

In 3.5 gamepads were working great through GamePad. All were recognised and I could plug them while in game and they would work with no issues.

@cra0zy @mrhelmut

**Comment by mrhelmut** [comment]
1. I will take a look at this, but I don't know when yet. It's on my to do list, I will need it to work too.

2. If gamepads are not recognized, it means that they are not in the "known gamepads" database. Feel free to add them by adding their GIUD and mapping to this file: https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework/SDL/gamecontrollerdb.txt

3. Should be fixed by #4949 (let me know if it works).

**Comment by endimilojkoski** [comment]
Yup, #4949 fixes the Mac game crash with gamepad plugged in.
So far latest version is working fine with no critical issues on all 3 platforms.

About #2, can't you somehow include the old OpenTK definitions or this is different?
Even if I add my 2 joysticks, how can I be sure that many other joysticks that I don't have will work for the players?

**Comment by endimilojkoski** [comment]
One of my gamepads that is not recognised is iBuffalo and entry already exists:

83056020000000000000504944564944,iBuffalo USB 2-axis 8-button Gamepad,a:b1,b:b0,y:b2,x:b3,start:b7,back:b6,leftshoulder:b4,rightshoulder:b5,leftx:a0,lefty:a1,platform:Windows,

I see that the ID is not matching so I copied the existing entry and changed the ID to match what I got from Joystick.GetCapabilities(1).Id:

20600583000000000000504944564944,iBuffalo USB 2-axis 8-button Gamepad,a:b1,b:b0,y:b2,x:b3,start:b7,back:b6,leftshoulder:b4,rightshoulder:b5,leftx:a0,lefty:a1,platform:Windows,

But it is still not being added to Gamepads and therefore not recognised in game.

**Comment by mrhelmut** [comment]
Thanks for testing #4949 !

> About 2, can't you somehow include the old OpenTK definitions or this is different?
Even if I add my 2 joysticks, how can I be sure that many other joysticks that I don't have will work for the players?

I'll try to make a diff with the old OpenTK database and your info to add these gamepads. iBuffalo gamepad look to be popular, I already saw quite a few being used.

**Comment by endimilojkoski** [comment]
OK, that would be great.
This is very important for my game, players must be able to play it with any gamepad like they can now.

And why is iBuffalo still not working for me after I add the correct ID?

Hope this gets improved quickly, it is holding back my update :)

**Comment by mrhelmut** [comment]
I pushed an update to the DB (#4954) but I can confirm that the iBuffalo entry is correct (GUID headers are read inverted, 20 60 05 83 is read 83 05 60 20, which is correct). OpenTK had the same DB, so I don't know what went wrong.

Let me know if you find why it doesn't work, I don't own an iBuffalo so I can't test it.

**Comment by endimilojkoski** [comment]
I got your latest changes and still it does not work.
It is not adding the gamepad here:
            while (Sdl.PeepEvents(events, 1, Sdl.EventAction.GetEvent, Sdl.EventType.ControllerDeviceAdded, Sdl.EventType.ControllerDeviceAdded) == 1)
            {
                GamePad.AddDevice(events[0].ControllerDevice.Which);
            }

What else can I test?

**Comment by mrhelmut** [comment]
My changes were not meant to fix the iBuffalo, and I haven't worked on the hot plug/unplug issue yet.

The only thing to do for the iBuffalo is to tweak the gamecontrollerdb.txt file until it is detected. I can't help without an iBuffalo unfortunately, so it's up to you to figure this out.

**Comment by endimilojkoski** [comment]
The thing is that my old Genius X360-like gamepad is also not recognised and both were working out of the box before through GamePad.
But they are both added to Joysticks and they work fine through Joystick.

**Comment by mrhelmut** [comment]
We prevented gamepads which are not in the DB from being considered as gamepads (this was not the case before) because there are a lot of gamepads on the market which lacks buttons / axes compared to what the GamePad class of MonoGame requires (otherwise there is a very high risk that they make MonoGame to crash whenever a missing button/axis is queried).
For this reason, we can't let them be added to GamePad unless their full mapping is specified in the controller DB. If they are not in the DB, they are added to Joystick in order to avoid crashes and let the mapping up to the developer.

If you wish them to be supported by GamePad, you will need to figure out what to add to the DB file and I'll gladly push an update. I can't do this alone, because I don't own any of these controllers.

**Comment by endimilojkoski** [comment]
Ahhh, this is so frustrating, have been changing my logic with GamePad/Joystick several times :)
I have never had any issues with GamePad before.

I just tried it on Linux. iBuffalo and the old Genius work with no issues with GamePad with OpenTK and they are both not working with latest SDL2 version.

**Comment by endimilojkoski** [comment]
Once I moved from GamePad to Joystick because I needed 8 gamepads support that GamePad was not offering.
But I started having game crashing issues with Joystick and once you changed GamePad to support more than 4 I switched back to it and it was all good.

**Comment by endimilojkoski** [comment]
I got you and I am not blaming you, just saying that it worked great for me so far and now that is not the case anymore and I have to work on sorting out input issues again.

And the fact that iBuffalo is in the DB and not working for me indicates that there are still issues.

**Comment by mrhelmut** [comment]
Issues are in the DB. This code path is not MonoGame related, controller support through this DB is purely a SDL thing. SDL is in charge of loading this DB and allowing controllers to support the GamePad API.

The only solution is to find what is wrong with the DB, and that I can not help without owning the controllers.

**Comment by endimilojkoski** [comment]
So what would you suggest me to do?
Is moving back to Joystick the only solution to make sure players can use any gamepad?

**Comment by cra0zy** [comment]
Steam has it's own SDL database which any SDL game should be able to tap into.

Now in general, 3rd party controllers won't have any support unless there is an entry in the database file, Xna didn't have any API for remapping so MonoGame doesn't have it either, but you should be able to load your custom SDL mapping configs with https://wiki.libsdl.org/SDL_GameControllerAddMapping?highlight=%28%5CbCategoryGameController%5Cb%29%7C%28CategoryEnum%29

**Comment by cra0zy** [comment]
PS. SDL mapping is a string which examples you can see in https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework/SDL/gamecontrollerdb.txt, you could use Joystick class to allow players to add their own config.

**Comment by mrhelmut** [comment]
> So what would you suggest me to do?
Is moving back to Joystick the only solution to make sure players can use any gamepad?

If you wish that every gamepad in the world works, yes, you will have to use the Joystick API and manage remapping in your game.

If you are fine with the DB and all the info provided by cra0zy, then go with the GamePad API.

Note that the DB covers 99% of the gamers (based on our reports).

**Comment by endimilojkoski** [comment]
OK, I see that but it means that I will have to do a lot of changes to make it work properly.

So you are saying I should utilise Joystick to enable players to set up their gamepads and to create new GamePad mapping entry?
I need a quick solution. Can I do something to add all Joysticks by default under GamePads with X360 mappings?

**Comment by cra0zy** [comment]
> I need a quick solution. Can I do something to add all Joysticks by default under GamePads with X360 mappings?

No.

**Comment by endimilojkoski** [comment]
@mrhelmut, 99% is fine with me for GamePad but iBuffalo which is in the mappings is not working and that makes me worried. Have you tested non X360 gamepads with NeuroVoider?

I just tried NV with my Genius X360-like gamepad and it is not recognized but for example TowerFall recognize it right away.
And iBuffalo is also not recognised in NV even though it is in the gamepad DB.

**Comment by tomspilman** [comment]
> And iBuffalo is also not recognised in NV even though it is in the gamepad DB.

That has to be a bug of some sort in the SDL implementation in MonoGame.

**Comment by cra0zy** [comment]
> That has to be a bug of some sort in the SDL implementation in MonoGame.

More likely a bug in the database file.

**Comment by mrhelmut** [comment]
+1 for the DB bug. The whole game controller mechanism happens inside SDL, if ```PeepEvents``` doesn't return ```GameControllerAdded``` events, it surely is the DB.

@endimilojkoski you should try editing the iBuffalo entry in the DB until it works, the GUID may be good but not in the good order.

**Comment by endimilojkoski** [comment]
Checked the mapping line and it all looks good.
Tried to rearrange the GUID but nothing changed.

Can some of you try any non X360 controller through GamePad?

**Comment by mrhelmut** [comment]
Everything is fine on my end with PS4, Logitech gamepads and Steam Controller.

Note that Steam replaces the DB with its own (and may not support iBuffalo). You should try without the Steam API.

**Comment by endimilojkoski** [comment]
Also tried without Steam api and Steam client turned off with no luck.
OK, I will go ahead and update the game but I hope very we find a solution for this.

Btw, forgot to mention that now I don't see any ghost devices under GamePad and Joystick as before. I added some of the latest changes you did and it probably fixed that.

**Comment by mrhelmut** [comment]
1. I've tested a lot of situations with different brands of gamepads, but I can't reproduce the crash.

However, I found a problem which prevents GamePad to be detected post-launch in ```SDLGamePlatform.SdlRunLoop()```, there's a ```else if (ev.Type == Sdl.EventType.ControllerDeviceAdded) GamePad.AddDevice(ev.ControllerDevice.Which);``` missing.

Would you mind trying to add it and tell me if it fixes 1. ? Make sure that your are using the very last develop branch and the SDL2.dll from here (because other versions may make events to fail): https://github.com/MonoGame/MonoGame.Dependencies/tree/a62d80be574ef74ee4196e3f7f1a40835634d857/SDL


With this fix, ```GamePad``` and ```Joystick``` are working totally fine for me with hot plug/unplug support.

**Comment by endimilojkoski** [comment]
Got latest dev source and got SDL2 from that location and still issues.
When I plug in the gamepad in game I get the same issue:
![image](https://cloud.githubusercontent.com/assets/13436441/16380984/7c091e18-3c7a-11e6-9964-3be6705b1b15.png)

And now I can see 1 additional ghost GamePad and Joystick device as before and it causes this issue, the connected GamePad is added in the dictionary with existing key.
I am not sure what fixed the ghost issue for me recently but it is happening again.

**Comment by endimilojkoski** [comment]
OK, I have good news.
I solved the Dictionary key bug by simply checking if the key exists, if it does I update the existing one.

I did this in Joystick.SDL.cs -> AddDevice:
![image](https://cloud.githubusercontent.com/assets/13436441/16383368/20beca66-3c85-11e6-907e-675ced897f75.png)

And in GamePad.SDL.cs -> AddDevice:
![image](https://cloud.githubusercontent.com/assets/13436441/16383389/3ad7ab7a-3c85-11e6-8dfc-dbdece86e668.png)

Now if I plug in the gamepad while in game it simply replaces the ghost key and it works fine.
I ahve tried this before and it was not working, I guess your code that I added in SdlRunLoop made this work.

And when I plug in my iBuffalo gamepad while in game it works!

Let me test some more and will let you know.

**Comment by mrhelmut** [comment]
You shouldn't have to do that, if there is a ghost id, it means that the device removal didn't went well, and that's an issue. By replacing a ghost, you're hiding the real issue, and it may have side effects.

This looks like a problem with ```deviceId``` always being 0, and it is often associated to SDL bindings (there was an error related to this which has been fixed). Please verify that you are using the latest develop branch and the linked SDL2.dll.

I believe everything is fixed with the proposed change.

**Comment by endimilojkoski** [comment]
Yup, confirmed. If I start the game with iBuffalo plugged in it is not recognised here:
![image](https://cloud.githubusercontent.com/assets/13436441/16384044/dce8cf96-3c87-11e6-9aba-d6aab9e950ef.png)

If I plug it in game or if I unplug it and plug it back it is recognised and working fine:
![image](https://cloud.githubusercontent.com/assets/13436441/16384194/5013d628-3c88-11e6-8f5d-b53b27a7a17e.png)

**Comment by endimilojkoski** [comment]
Yes, I got latest develop version and the SDL from the link you sent me.

**Comment by mrhelmut** [comment]
That is very strange, but unrelated to 1.

**Comment by endimilojkoski** [comment]
Yes, and recently this issue went away and is now back again.
What else can I test and try to solve this?

**Comment by endimilojkoski** [comment]
I am using 32-bit SDL2 if that makes any difference, forced 32-bit only version of the game.

**Comment by mrhelmut** [comment]
Is it only related to the iBuffalo? Do you have any crash with X360 gamepads? (without your modifications, to see it hides something)

**Comment by endimilojkoski** [comment]
X360 gamepad is working fine if I start the game with it but if I plug it in while in game it crashes due to the existing Dictionary key from the ghost device.
With my changes X360 gamepad is working fine in all cases.

**Comment by mrhelmut** [comment]
Ok, thanks.
I really don't have any problem, and I don't think that your changes are a safe solution because it hides the ghost device id issue, which need to be fixed first.

When you plug a X360 gamepad while in game, can you check all the value ```ev.ControllerDevice.Timestamp```? Is it 0?

**Comment by endimilojkoski** [comment]
![image](https://cloud.githubusercontent.com/assets/13436441/16386509/33b5dc52-3c91-11e6-84e0-3015e3332afc.png)

**Comment by mrhelmut** [comment]
Thanks, it looks ok.

I don't know what is going on, it increments / decrements correctly on my side. I tested this with up to 6 controllers of different brands.

Can you try this inside of ```GamePad.AddDevice()```? (and the equivalent in Joystick)
```
            if (Gamepads.ContainsKey(deviceId) && Gamepads[deviceId].Device == gamepad.Device)
                Gamepads.Remove(deviceId);
            Gamepads.Add(deviceId, gamepad);
```

**Comment by endimilojkoski** [comment]
This is not equal && `Gamepads[deviceId].Device == gamepad.Device` and it fails again.
If I remove it from the condition it removes the Gamepad and the Joystick successfully and after adding it it also works fine.

**Comment by mrhelmut** [comment]
I see two possibilities then:
1. The ghost device hasn't been correctly cleaned when unplugged (did you unplugged anything before trying to (re)plug something else?);
2. The problem doesn't happen after unplugging anything, and it would then be something wrong with how we handle SDL devices.

**Comment by endimilojkoski** [comment]
The ghost devices are there since the start of the game without anything plugged in and has nothing to do with plugging/unplugging.

**Comment by endimilojkoski** [comment]
I have nothing plugged in:
![image](https://cloud.githubusercontent.com/assets/13436441/16389199/22c2a810-3c9d-11e6-8908-572b26053e53.png)

**Comment by mrhelmut** [comment]
```BeforeInitialize()``` does its normal job, I don't see any problem there.

What interest me is the content of ```GamePad.GamePads``` when ```else if (ev.Type == Sdl.EventType.ControllerDeviceAdded) GamePad.AddDevice(ev.ControllerDevice.Which);``` is reached. Also, what devices do you have plugged in when starting the game, and what device are you trying to hot plug then.

**Comment by endimilojkoski** [comment]
The screenshot above is showing the ghost device, I have nothing connected.

**Comment by tomspilman** [comment]
>  I have nothing connected.

Might try unhooking the mouse, speakers, or any other external device even if it isn't an input device.  Could be something is being misidentified.

**Comment by mrhelmut** [comment]
Something crosses my mind. In ```BeforeInitialize```, we peep device addition, but never device removal. If SDL gives you an ID of 0 during gameplay, it means that it removed this ghost device somewhere, most probably in ```BeforeInitialize```.

What if you had this to ```BeforeInitialize```?
```
			while (Sdl.PeepEvents(events, 1, Sdl.EventAction.GetEvent, Sdl.EventType.ControllerDeviceRemoved, Sdl.EventType.ControllerDeviceRemoved) == 1)
            {
                GamePad.RemoveDevice(events[0].ControllerDevice.Which);
            }
            while (Sdl.PeepEvents(events, 1, Sdl.EventAction.GetEvent, Sdl.EventType.JoyDeviceRemoved, Sdl.EventType.JoyDeviceRemoved) == 1)
            {
                Joystick.RemoveDevice(events[0].JoystickDevice.Which);
            }
```

Is there any event caught by those lines?

**Comment by cra0zy** [comment]
I remember that I put in `SdlRunLoop();` in BeforeInitalize... Why was it replaced again?

**Comment by endimilojkoski** [comment]
@tomspilman I have disconnected everything from my laptop and it still adds to GamePad and Joystick.

@mrhelmut I added those for test and they do nothing.

**Comment by mrhelmut** [comment]
@cra0zy I believe because gamepads were not populated until ```Game.Update()```.

If this ghost device lands in ```GamePad```, it means that its GUID is in the controller database. It MUST be something from this list, or anything having the same GUID as something else.

You should try using ```SDL_JoystickGetGUID``` or ```SDL_JoystickGetDeviceGUID``` to get its GUID in ```AddDevice``` and identify it in the list.

I wouldn't be surprised that it is related to a wireless receiver or a USB hub.

**Comment by cra0zy** [comment]
On the case of ghost device, turns out PS 4 controller was getting detected even tho it was turned off. The indexes were still tangled, I guess I'll add a better way of handling devices.

On the case of why the IBufallo device not working, turns out that ControllerAdded event wasn't getting triggered by SDL, but doing a IsAGameController when JoystickEvent gets triggered did work, so we have a good workaround.

I'll submit a PR on this a bit later today.

**Comment by mrhelmut** [comment]
Cool funding.
Do you mind cherry picking this? (if it is relevant to your changes) https://github.com/mrhelmut/MonoGame/commit/a36c99821e6b3eb300fe23a490a864a02271d13c#diff-3f1604aca3e5dd0a42ffb436245dcad9R129

**Comment by cra0zy** [comment]
`Sdl.EventType.ControllerDeviceAdded` doesn't always get triggered... so that commit won't do anything useful.

**Comment by endimilojkoski** [comment]
With @cra0zy 's changes now GamePad is working as it should. Great job guys!
I am closing this now.
